rlclientlib/federation/sender_joined_log_provider.cc

```diff
@@ -0,0 +1,222 @@
+  output.insert(std::end(output), reinterpret_cast<uint8_t*>(&size), reinterpret_cast<uint8_t*>(&size) + 4);
+
+  output.reserve(output.size() + size);
+  output.insert(std::end(output), buffer.data(), buffer.data() + size);
+  return 0;
+}
+}  // namespace
+
+namespace reinforcement_learning
+{
+int sender_joined_log_provider::init(
+    const reinforcement_learning::utility::configuration& config, reinforcement_learning::api_status* status)
+{
+  if (config.get_int(reinforcement_learning::name::PROTOCOL_VERSION, 999) != 2)
+  { RETURN_ERROR_LS(_trace_logger, status, invalid_argument) << " protocol version 2 required"; }
+
+  const auto* eud_duration = config.get(reinforcement_learning::name::EUD_DURATION, "UNSET");
+  if (eud_duration == reinforcement_learning::string_view("UNSET"))
+  { RETURN_ERROR_ARG(_trace_logger, status, invalid_argument, "eudduration must be set"); }
+  RETURN_IF_FAIL(reinforcement_learning::parse_eud(eud_duration, _eud_offset, status));
+  return error_code::success;
+}
+
+RL_ATTR(nodiscard)
+int sender_joined_log_provider::invoke_join(std::unique_ptr<VW::io::reader>& batch, api_status* status)
+{
+  std::lock_guard<std::mutex> lock(_mutex);
+  std::vector<uint8_t> result;
+  const auto eud_cutoff = std::chrono::system_clock::now() - _eud_offset;
+  emit_filemagic_message(result);
+  for (auto it = _interactions.cbegin(); it != _interactions.cend();)
+  {
+    flatbuffers::FlatBufferBuilder fbb;
+    std::vector<flatbuffers::Offset<reinforcement_learning::messages::flatbuff::v2::JoinedEvent>> events;
+
+    const auto& item = *it;
+    const auto interaction_time = chrono_from_timestamp(std::get<0>(it->first));
+    if (interaction_time <= eud_cutoff)
+    {
+      const auto reward_cutoff = interaction_time + _eud_offset;
+
+      auto ts = from_rl_timestamp(std::get<0>(it->first));
+      events.push_back(reinforcement_learning::messages::flatbuff::v2::CreateJoinedEventDirect(fbb, &it->second, &ts));
+      const auto event_id = std::get<1>(it->first);
+      const auto& observations = _observations.find(event_id);
+      if (observations != _observations.end())
+      {
+        for (const auto& observation : observations->second)
+        {
+          auto observation_time = chrono_from_timestamp(std::get<0>(observation));
+          if (observation_time <= reward_cutoff)
+          {
+            auto ts = from_rl_timestamp(std::get<0>(observation));
+            events.push_back(reinforcement_learning::messages::flatbuff::v2::CreateJoinedEventDirect(
+                fbb, &std::get<1>(observation), &ts));
+          }
+        }
+        _observations.erase(event_id);
+      }
+      _interactions.erase(it++);
+    }
+    else
+    {
+      // TODO make sure this is correct. As in does the timestamp ordering mean that all further events are within eud.
+      // break;
+      ++it;
+    }
+
+    auto joined_payload = reinforcement_learning::messages::flatbuff::v2::CreateJoinedPayloadDirect(fbb, &events);
```

Comment:
should this part go inside the if statement?

Reply:
Good catch definitely should